Add Figure 5.6

j-zimmermann opened the pull request:
<!-- To ensure we can review your pull request promptly please complete this template entirely. -->

Dear authors,

I enjoyed reading your book; thank you for your work and particularly for sharing so much material with your readers!

In chapter 5, you have discussed a paper of which I am the first author (Zimmermann & van Rienen (2021)).
Because I saw that you read the values of our model visually off, I thought it might be of value to share the code behind the model.
Here, I have implemented the comparison between the Gabriel model and our model for gray matter (Fig. 5.6) and white matter (not shown in the book).
I have not added the data from Wagner et al.

Please let me know if I you would like to include this notebook in your collection and if I can change anything.
I have tried to follow your instructions (e.g., I deleted the output of all notebook cells).

Thank you very much for your feedback,
Julius

<!-- Please reference the issue number here. You can replace "Fixes" with "Closes" if it makes more sense. -->

No related issue.

Changes proposed in this pull request:
<!-- Please list all changes/additions here. -->
- I added Figure 5.6 (the formatting is not exactly the same as in print).

## Before submitting

<!-- Please complete this checklist BEFORE submitting your PR to speed along the review process. -->
- [x] I've read and followed all steps in the [Making a pull request](https://github.com/LFPy/ElectricBrainSignals/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [ ] I've bumped the version number in `brainsignals/version.py` file
- [ ] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/LFPy/ElectricBrainSignals/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [ ] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

j-zimmermann committed:
add new figure

espenhgn commented:
Thanks, @j-zimmermann; this is really nice! 
And thank you for the nice words on our book, I hope you found it useful :)
I don't have the Wagner data on hand, but I can check with the others here if we can easily add it. 

Otherwise, I'm more than happy to merge as is, as this is still a useful resource to include here. 
I could run the notebook without issues on my end.

espenhgn commented:
For future reference, these were the Wagner data (from @torbjone):
```
# [Frequency (Hz), Conductivity (S/m), Relative Permittivity]
# From Wagner et al. (2014) Neuroimage 85:1048–1057.

import numpy as np
wagner_data = np.array(
    [[10.0000, 0.32080, 31630000.0],
    [11.2200, 0.32170, 29940000.0],
    [12.5890, 0.32860, 28450000.0],
    [14.1250, 0.32250, 27500000.0],
    [15.8480, 0.32240, 26450000.0],
    [17.7820, 0.32560, 25120000.0],
    [19.9520, 0.32720, 23800000.0],
    [22.3870, 0.32770, 22400000.0],
    [25.1180, 0.33360, 21290000.0],
    [28.1830, 0.33370, 21280000.0],
    [31.6220, 0.33650, 21270000.0],
    [35.4810, 0.33770, 21500000.0],
    [39.8100, 0.33850, 21730000.0],
    [44.6680, 0.34310, 21230000.0],
    [50.1180, 0.35760, 20930000.0],
    [56.2340, 0.35860, 19880000.0],
    [63.0950, 0.35910, 19200000.0],
    [70.7940, 0.33610, 18490000.0],
    [79.4320, 0.34830, 17590000.0],
    [89.1250, 0.33510, 16240000.0],
    [100.0000, 0.33360, 15280000.0],
    [112.2010, 0.33470, 14650000.0],
    [125.8920, 0.33930, 13870000.0],
    [141.2530, 0.34270, 13090000.0],
    [158.4890, 0.34320, 12260000.0],
    [177.8270, 0.34580, 11260000.0],
    [199.5260, 0.34460, 10560000.0],
    [223.8720, 0.34690, 9720000.0],
    [251.1880, 0.35180, 9293000.0],
    [281.8380, 0.35320, 8774000.0],
    [316.2270, 0.35410, 8316000.0],
    [354.8130, 0.35710, 7746000.0],
    [398.1070, 0.35730, 7187000.0],
    [446.6830, 0.35850, 6743000.0],
    [501.1870, 0.35760, 6177000.0],
    [562.3410, 0.36010, 5790000.0],
    [630.9570, 0.36360, 5394000.0],
    [707.9450, 0.37360, 5110000.0],
    [794.3280, 0.38790, 4796000.0],
    [891.2500, 0.39640, 4408000.0],
    [1000.0000, 0.40260, 4091000.0],
    [1122.0180, 0.40850, 3724000.0],
    [1258.9250, 0.41040, 3433000.0],
    [1412.5370, 0.41440, 3104000.0],
    [1584.8930, 0.42740, 2951000.0],
    [1778.2790, 0.43900, 2700000.0],
    [1995.2620, 0.46540, 2592000.0],
    [2238.7210, 0.48410, 2398000.0],
    [2511.8860, 0.51010, 2128000.0],
    [2818.3820, 0.54550, 1939000.0],
    [3162.2770, 0.54860, 1718000.0],
    [3548.1330, 0.56360, 1593000.0],
    [3981.0710, 0.57130, 1419000.0],
    [4466.8350, 0.59510, 1317000.0],
    [5011.8720, 0.61280, 1192000.0],
    [5623.4130, 0.63200, 1064000.0],
    [6309.5730, 0.64270, 944100.0],
    [7079.4570, 0.64220, 819700.0],
    [7943.2820, 0.65870, 732100.0],
    [8912.5090, 0.63270, 661400.0],
    [10000.0000, 0.63350, 607100.0],
    [11220.1800, 0.62270, 464900.0],
    [12589.2500, 0.62940, 405400.0],
    [14125.3700, 0.63830, 355800.0],
    [15848.9300, 0.65710, 312100.0],
    [17782.7900, 0.64870, 263400.0],
    [19952.6200, 0.66910, 232800.0],
    [22387.2100, 0.66730, 197800.0],
    [25118.8600, 0.69100, 176700.0],
    [28183.8200, 0.70110, 152100.0],
    [31622.7700, 0.73000, 136900.0],
    [35481.3300, 0.74250, 117700.0],
    [39810.7100, 0.76340, 104800.0],
    [44668.3500, 0.76660, 88760.0],
    [50118.7200, 0.76630, 77490.0]])


f = wagner_data[:, 0]
sigma = wagner_data[:, 1]
permittivity = wagner_data[:, 2]
plt.plot(f, sigma)
plt.show()
```

j-zimmermann committed:
update notebook with Wagner data

j-zimmermann commented:
Thank you very much @espenhgn ! 

I updated the notebook with the Wagner data and added another plot that compares the Wagner data to the models over the entire frequency range of the Wagner data.

Let me know if there's anything else I can do.

espenhgn commented:
Thanks a lot for the update! 
I did however notice a couple of differences between our figure and this implementation.
In panel B from the book the "Zimmermann" trace starts at a value of 1E7 at 10Hz, but here it is at 2E7; in panel C there's a ripple in the Gabriel trace just below 10kHz, which is not reproduced here. 
There are some minute differences across panels for the Zimmermann traces in panel A and C, mainly a bump/increase in values near 10kHz. 

Do you reckon there's an issue with what we produced for the book? 
Note that I have not compared the codes accordingly - @geih made the plot originally. 

<img width="690" alt="image" src="https://github.com/user-attachments/assets/1ee66fc5-ed97-4f4a-91c1-acb1c81ee333" />

![image](https://github.com/user-attachments/assets/4bb58422-1bc4-48c4-acda-1e38f105b975)

j-zimmermann committed:
Update Figure 5.6 with HN model

j-zimmermann commented:
Hi @espenhgn, Thanks for pointing out the deviations. I had a look at my code and the data and double-checked the values with the ITIS database.

* The difference in Panel B  comes from the fact that you have plotted the results of our fit of the Gabriel data to the Havriliak-Nagami model (HN model). I did not notice that when reading because we usually use the 3-Cole-Cole (3CC) model. We observed that the HN model is usually in very good agreement with the 3CC model, which is the original Gabriel model with an adjusted static conductivity. We took the static conductivity from the fitted HN model parameters. While the relative permittivity of the HN model starts at 1e7, the relative permittivity of the 3CC model starts at 2e7.  The 2*pi*epsilon / sigma ratio is almost not affected by this difference because it is numerically close to zero at low frequencies. We usually prefer to use the 3CC model because it is easier to compare to the original Gabriel model. The plot in the book is not incorrect and I have added the HN model and its parameters to the notebook. Which version do you prefer?
* The minute differences in panel A and C can also be explained by the differences between the HN  and 3CC model.
* The ripple in the Gabriel trace does not make sense to me. I  double-checked the parameter values of the Gabriel model with the values used by the ITIS database and extracted the values at the frequencies between 7 and 8 kHz from the ITIS database. I did not observe a ripple, but I also have not tested all frequencies around the ripple. From the shape of the dielectric properties as described by the Gabriel model, I would not expect to observe a ripple. Would you prefer to have the tabulated Gabriel data in the notebook to compare with the function that I implemented?

espenhgn commented:
Hi Julius,
I now received the corresponding matlab codes from Geir  below (GitHub doesn't allow .m files to be uploaded). 

In short, the Gabriel data were from: https://itis.swiss/virtual-population/tissue-properties/database/dielectric-properties/

He entered chosen frequencies and read values for gray matter, obtaining:
```
% f, epsilon_r, sigma
Gfps = [10, 4.07e7, 2.75e-2
    20, 3.14e7, 4.3e-2
    40, 1.6e7, 6.8e-2
    60, 9.2e6, 8e-2
    80, 5.7e6, 8.6e-2
    100, 3.9e6, 8.9e-2
    150, 1.9e6, 9.2e-2
    200, 1.2e6, 9.4e-2
    300, 6.2e5, 9.5e-2
    500, 3.2e5, 9.6e-2
    750, 2.1e5, 9.8e-2
    1000, 1.6e5, 9.9e-2
    2000, 9.4e4, 1e-1
    3000, 6.7e4, 1.06e-1
    4000, 5.2e4, 1.08e-1
    5000, 4.2e4, 1.11e-1
    7500, 2.9e-4, 1.13e-1
    9000, 2.45e4, 1.14e-1
    10000, 2.22e4,1.15e-1]; 
```
Your data was read visually from a figure and smoothed with a spline, which may be inaccurate, obtaining:
```
Zf = 10:1:10000;
Zesamp = [10 100 1000 5000 10000];
Zssamp = [10 100 1000 3000 10000];
Zeps = [1e5 0.99e5 0.9e5 0.6e5 0.3e5]
Zsig = [0.11 0.11 0.11 0.115 0.14]
Zeps = spline(Zesamp, Zeps, Zf);
Zsig = spline(Zssamp, Zsig, Zf);

ep0 = 8.85e-12;
Zrat = 2*pi.*Zf*ep0.*Zeps./Zsig;
```
This may be the reason for the ripple artifact, as he read some values across a short interval in a log-log plot. 

I think having both HN and 3CC is useful for comparison and as a resource, and I can merge this as is unless you have something you would change still :)

```
% Wagnerdata4_en.m
%%% Zimmermann data
Zf = 10:1:10000;
Zesamp = [10 100 1000 5000 10000]; 
Zssamp = [10 100 1000 3000 10000]; 
Zeps = [1e5 0.99e5 0.9e5 0.6e5 0.3e5]
Zsig = [0.11 0.11 0.11 0.115 0.14]
Zeps = spline(Zesamp, Zeps, Zf); 
Zsig = spline(Zssamp, Zsig, Zf); 

ep0 = 8.85e-12; 
Zrat = 2*pi.*Zf*ep0.*Zeps./Zsig;

%%%% GABRIEL DATA
% f, epsilon_r, sigma
% Numbers for grey matter from 
% https://itis.swiss/virtual-population/tissue-properties/database/dielectric-properties/
Gfps = [10, 4.07e7, 2.75e-2
    20, 3.14e7, 4.3e-2
    40, 1.6e7, 6.8e-2
    60, 9.2e6, 8e-2
    80, 5.7e6, 8.6e-2
    100, 3.9e6, 8.9e-2
    150, 1.9e6, 9.2e-2
    200, 1.2e6, 9.4e-2
    300, 6.2e5, 9.5e-2
    500, 3.2e5, 9.6e-2
    750, 2.1e5, 9.8e-2
    1000, 1.6e5, 9.9e-2
    2000, 9.4e4, 1e-1
    3000, 6.7e4, 1.06e-1
    4000, 5.2e4, 1.08e-1
    5000, 4.2e4, 1.11e-1
    7500, 2.9e-4, 1.13e-1
    9000, 2.45e4, 1.14e-1
    10000, 2.22e4,1.15e-1]; 

fG = Gfps(:,1); 
sigmaG = Gfps(:,3);
epsG = Gfps(:,2); 
fGnew = 10:1:10000; 
sigmaGnew = spline(fG,sigmaG,fGnew); 
epsGnew = spline(fG,epsG,fGnew); 

ep0 = 8.85e-12; 
ratioG = 2*pi.*fG*ep0.*epsG./sigmaG;
ratioGnew = 2*pi.*fGnew*ep0.*epsGnew./sigmaGnew;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Wagne rdata
wagner_data = [10.0000, 0.32080, 31630000.0;
11.2200, 0.32170, 29940000.0;
12.5890, 0.32860, 28450000.0;
14.1250, 0.32250, 27500000.0;
15.8480, 0.32240, 26450000.0;
17.7820, 0.32560, 25120000.0;
19.9520, 0.32720, 23800000.0;
22.3870, 0.32770, 22400000.0;
25.1180, 0.33360, 21290000.0;
28.1830, 0.33370, 21280000.0;
31.6220, 0.33650, 21270000.0;
35.4810, 0.33770, 21500000.0;
39.8100, 0.33850, 21730000.0;
44.6680, 0.34310, 21230000.0;
50.1180, 0.35760, 20930000.0;
56.2340, 0.35860, 19880000.0;
63.0950, 0.35910, 19200000.0;
70.7940, 0.33610, 18490000.0;
79.4320, 0.34830, 17590000.0;
89.1250, 0.33510, 16240000.0;
100.0000, 0.33360, 15280000.0;
112.2010, 0.33470, 14650000.0;
125.8920, 0.33930, 13870000.0;
141.2530, 0.34270, 13090000.0;
158.4890, 0.34320, 12260000.0;
177.8270, 0.34580, 11260000.0;
199.5260, 0.34460, 10560000.0;
223.8720, 0.34690, 9720000.0;
251.1880, 0.35180, 9293000.0;
281.8380, 0.35320, 8774000.0;
316.2270, 0.35410, 8316000.0;
354.8130, 0.35710, 7746000.0;
398.1070, 0.35730, 7187000.0;
446.6830, 0.35850, 6743000.0;
501.1870, 0.35760, 6177000.0;
562.3410, 0.36010, 5790000.0;
630.9570, 0.36360, 5394000.0;
707.9450, 0.37360, 5110000.0;
794.3280, 0.38790, 4796000.0;
891.2500, 0.39640, 4408000.0;
1000.0000, 0.40260, 4091000.0;
1122.0180, 0.40850, 3724000.0;
1258.9250, 0.41040, 3433000.0;
1412.5370, 0.41440, 3104000.0;
1584.8930, 0.42740, 2951000.0;
1778.2790, 0.43900, 2700000.0;
1995.2620, 0.46540, 2592000.0;
2238.7210, 0.48410, 2398000.0;
2511.8860, 0.51010, 2128000.0;
2818.3820, 0.54550, 1939000.0;
3162.2770, 0.54860, 1718000.0;
3548.1330, 0.56360, 1593000.0;
3981.0710, 0.57130, 1419000.0;
4466.8350, 0.59510, 1317000.0;
5011.8720, 0.61280, 1192000.0;
5623.4130, 0.63200, 1064000.0;
6309.5730, 0.64270, 944100.0;
7079.4570, 0.64220, 819700.0;
7943.2820, 0.65870, 732100.0;
8912.5090, 0.63270, 661400.0;
10000.0000, 0.63350, 607100.0;
11220.1800, 0.62270, 464900.0;
12589.2500, 0.62940, 405400.0;
14125.3700, 0.63830, 355800.0;
15848.9300, 0.65710, 312100.0;
17782.7900, 0.64870, 263400.0;
19952.6200, 0.66910, 232800.0;
22387.2100, 0.66730, 197800.0;
25118.8600, 0.69100, 176700.0;
28183.8200, 0.70110, 152100.0;
31622.7700, 0.73000, 136900.0;
35481.3300, 0.74250, 117700.0;
39810.7100, 0.76340, 104800.0;
44668.3500, 0.76660, 88760.0;
50118.7200, 0.76630, 77490.0];

fw = wagner_data(:,1); 
sigmaw = wagner_data(:,2); 
epsilonw = wagner_data(:,3); 


% %% WAGNER DATA
% sigma = [10; 0.3111111111111111
% 87.03856749311285; 0.3497053872053872
% 200; 0.346
% 406.19834710743794; 0.3457070707070708
% 725.358126721763; 0.3583754208754209
% 1055.5234159779613; 0.3876683501683502
% 1418.7052341597796; 0.39461279461279464
% 1781.8870523415976; 0.41544612794612795
% 2420.206611570248; 0.47411616161616166
% 2827.410468319559; 0.5142255892255893
% 3212.6033057851237; 0.5183080808080809
% 3619.8071625344355; 0.5306397306397307
% 4137.06611570248; 0.5397727272727273
% 4918.457300275481; 0.5701178451178451
% 5512.75482093664; 0.5872895622895623
% 6008.002754820937; 0.596506734006734
% 6646.322314049587; 0.5940656565656566
% 7130.564738292011; 0.5949915824915826
% 7625.812672176309; 0.6042087542087542
% 8088.044077134986; 0.6024410774410774
% 8660.330578512398; 0.5835858585858587
% 9331.666666666666; 0.5699074074074074
% 9969.986225895316; 0.5730218855218856]; 
% 
% epsilon = [10.0; 2.3e7
% 231; 1e7
% 420; 7e6
% 520; 6e6; 
% 680; 5e6
% 1000; 4e6
% 1400; 3e6
% 2700; 2e6
% 6000; 1e6
% 6700; 9e5
% 7600; 8e5;
% 9000; 7e5; 
% 10000; 6e5]; 

% sx = sigma(1:2:end-1);
% sy = sigma(2:2:end);
% [sxs Ix] = sort(sx);
% sx = sx(Ix);
% sy = sy(Ix); 
% x = [10:1:10000];
% %[x, Ix] = x>10; 
% %sy = interp1(sx,sy,x); 
% sy = spline(sx,sy,x); 
% 
% epx = epsilon(1:2:end-1);
% epy = epsilon(2:2:end);
% mychoice = 3; 
% if mychoice == 1
%     epy = interp1(epx,epy,x);
% end
% if mychoice == 2
%     epy = spline(epx,epy,x);
% end
% if mychoice == 3 
%     lepy = log(epy); 
%     lepy = spline(epx,lepy,x); 
%     epy = exp(lepy); 
% end


ep0 = 8.85e-12; 
ratiow = 2*pi.*fw*ep0.*epsilonw./sigmaw;
figure(1); 
a = subplot(3,1,1)
semilogx(fw,sigmaw, 'k', 'LineWidth', 2); hold on; 
semilogx(fGnew,sigmaGnew,'r', 'LineWidth', 2)
semilogx(Zf, Zsig,'b', 'LineWidth',2); 

title('\sigma_{Rt}','FontName', 'FixedWidth')
ylabel('S/m'); 
set(a,'fontsize', 20)
axis([10 10000 0.025 0.7]);
%set(a, 'LineWidth', 2); 

a = subplot(3,1,2)
%plot(x,log10(epy), 'k', 'LineWidth', 2);
semilogx(fw,epsilonw, 'k', 'LineWidth', 2); hold on; 
semilogx(fGnew, epsGnew,'r', 'LineWidth',2); 
semilogx(Zf, 100*Zeps,'b', 'LineWidth',2); 

title('\epsilon_{rt}','FontName', 'FixedWidth')
set(a,'fontsize', 20)
legend('Wagner','Gabriel','Zimmermann')
%set(a, 'LineWidth', 2); 
haxis= axis; 
axis([10 10000 haxis(3) haxis(4)]);


a = subplot(3,1,3)
semilogx(fw,ratiow,'k', 'LineWidth', 2); hold on; 
semilogx(fGnew, ratioGnew,'r', 'LineWidth',2); 
semilogx(Zf, Zrat,'b', 'LineWidth',2); 

title('2{\pi}f\epsilon_{t}/\sigma_{Rt}', 'FontName', 'FixedWidth')
xlabel('f(Hz)')
set(a,'fontsize', 20)
%set(a, 'LineWidth', 2); 
haxis= axis; 
axis([10 10000 haxis(3) haxis(4)]);


if 0
figure(1); hold on; 
subplot(3,1,1)
plot(fGnew, sigmaGnew,'r'); 
subplot(3,1,2)
plot(fGnew, epsGnew,'r'); 
subplot(3,1,3); 
plot(fGnew, ratioG,'r'); 
end
```

j-zimmermann commented:
Thanks for merging, it looks alright to me!
Thank you also for your replies and sharing all the code and data!
